### PR TITLE
Simplify autodiff backend logic

### PR DESCRIPTION
**Pull request by nkoep**
This PR adds a first implementation of the changes suggested in #55. It's based on my previous PR (#56), and also adds the necessary changes to resolve the remaining issues outlined in the original issue. Refer to the included examples to see how exactly the proposed API works now. Note that this PR also includes my earlier experimental pytorch backend which ideally should be combined with / replaced by #66 down the line.

Due to the invasive nature of these changes, I figured we should open up a development branch to see how things develop with the new backend logic before merging this into master. If nothing else, we need to create a tag/release before merging this PR into master so early users of pymanopt have a reference version to go back to once we broke their code.

### CODE REVIEW

**Comment by nkoep** [comment]
I had another idea to potentially normalize the look and feel of different cost functions/gradients/Hessians. In particular, one thing I don't like about the current implementation after thinking about it a bit more is that we are really abusing python's decorator syntax. This is what I had in mind instead:

```python
from pymanopt import Autograd, Theano

@Autograd()
def cost(x):
    return np.sum(x ** 2)


X = T.vector()

@Theano(X)
def gradient(X):
    return 2 * X
```

Instead of having generic wrapper classes for each backend as defined in https://github.com/pymanopt/pymanopt/blob/43afe04616fe1b8ab8b1f5d1e03668bec31fee19/pymanopt/tools/_functions.py#L5 we would instead write proper functions to be used as decorators. For tracing backends such as autograd and pytorch these would simply wrap the callable in the respective *Function class as in the current implementation. For graph-based backends like theano and tensorflow, however, the decorators would first "unwrap" the callables by calling the functions with the respective symbolic variable once to construct the graph, before passing it on internally to their *Function counterparts.

I'm in the process of hacking up a solution for this proposal and noticed another advantage. Since we have to support multiple arguments in the decorators for product manifolds, the individual `arg` argument will always be an iterable, i.e., a tuple. This means that for differentiation and compilation purposes in the backends, we only have to consider the respective product manifold paths. I'm gonna continue working on this to see if I hit a roadblock somewhere down the line, but so far I'm optimistic.

**Comment by sweichwald** [comment]
Just stumbled again upon this proposed efficiency improvement from #65, which is related to this rewrite of the backend logic.

**Comment by calincru** [comment]
As I can tell from https://github.com/pymanopt/pymanopt/pull/66#issuecomment-502156232, this has not been forgotten/dropped.  Am I right?  Really looking forward to it, thanks for the work you've put in.

**Comment by nkoep** [comment]
Alright, that was quite a sprint! I've just pushed the last few commits towards reworking the autodiff backend implementation. At this point I guess the name I chose for this PR seems somewhat egregious ;D

There's still a little clean-up to do here and there (such as variable naming in the pytorch backend, factoring out code repetition in the unit tests, and eliminating redundancy in certain backends), but these issues could also be addressed by specialized PRs.

Highlights of this PR include:
- Support product manifolds, manifolds with factored point representations, as well as mixtures of the two in every backend (see also https://github.com/pymanopt/pymanopt/issues/23)
- Allow users to pass in their own tensorflow sessions (see https://github.com/pymanopt/pymanopt/issues/59, https://github.com/pymanopt/pymanopt/issues/57, https://github.com/pymanopt/pymanopt/pull/58)
- Add PyTorch autodiff backend (https://github.com/pymanopt/pymanopt/issues/52)

The current implementation of the tensorflow backend would also allow for a relatively straightforward way for users to pass in custom `feed_dicts`, although this would currently be limited to `feed_dicts` which are ready by the time the cost function is defined. I have some ideas on how to work around this limitation though if necessary. Maybe @vinay-jayaram and @luk-f-a can chime in on this.

The PyTorch backend currently does not integrate the caching approach proposed by https://github.com/pymanopt/pymanopt/pull/66. We could think about adopting the approach later on should the need arise, but for now I'd prefer to leave it out.

Admittedly, I was fighting PyTorch all the way while implementing the backend since I'm not deeply familiar with its best practices, so I'm sure there are more elegant ways to perform certain operations. The `compute_hessian` method for functions of multiple varibles seems particularly lengthy at the moment. One thing I was bitten by is that there seems to be no way to force PyTorch to retain gradients if they're independent/unused rather than replacing them with `None`. This happens for instance when computing Hessian-vector products of functions with linear components. In those cases it seems necessary to manually return a zero tensor of appropriate size. If anyone knows of a more elegant solution, let me know.

For now, I think I'm off for a pint :beer:

**Comment by nkoep** [comment]
Re passing in custom `feed_dicts`, I just realized that this can now easily be achieved by subclassing `tf.Session` along the following lines:
```python
import tensorflow as tf


class Session(tf.Session):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._feed_dict = None

    def set_feed_dict(self, feed_dict):
        self._feed_dict = feed_dict

    def run(self, fetches, feed_dict=None, options=None, run_metadata=None):
        if feed_dict is None:
            feed_dict = self._feed_dict
        elif self._feed_dict is not None:
            feed_dict.update(self._feed_dict)
        super().run(fetches, feed_dict, options, run_metadata)
```
This means that we do not require any more changes in the backend at all to support user-provided `feed_dicts`.

**Comment by sweichwald** [comment]
Wow, @nkoep, what a sprint! :+1: :tada: 
Looking forward to reviewing this during the next couple of days; very solid from what I can tell so far!

**Comment by luk-f-a** [comment]
great to see all this work put into refreshing pymanopt!
